### PR TITLE
Configure and Start containerd phase

### PR DESCRIPTION
**Pull request by srm6867**
### Converting the containerd configure and start phases to golang. 

packaging the containerd and runc files with package.

**1)configure containerd**  

_start =>_ 
checks installed containerd version.
if not installed errors out.
if installed, configures the containerd config file, i.e `/etc/containerd/config.toml`

_stop and status_ => does nothing

**2)start containerd**

_start =>_ 
starts containerd service

_stop =>_
destroys containers in k8s and moby namespaces and stops the service

_status =>_
checks if containerd is active or not.

### CODE REVIEW

File: nodelet/pkg/phases/container_runtime/start_containerd.go [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+	if err != nil {
+		cp.log.Errorf("could not destroy containers in namespace: %s :%v", constants.K8sNamespace, err)
+		phaseutils.SetHostStatus(cp.hostPhase, constants.FailedState, err.Error())
+		return errors.Wrapf(err, "could not destroy containers in namespace: %s", constants.K8sNamespace)
+	}
+
+	// Stop the containerd service
+	cp.log.Infof("Stopping containerd")
+
+	exitCode, err = cp.cmd.RunCommand(ctx, nil, 0, "", "/bin/sudo", "/usr/bin/systemctl", "stop", constants.RuntimeContainerd)
```
Comment (pacharya-pf9) [inline nodelet/pkg/phases/container_runtime/start_containerd.go:110]: To keep things more readable implement 3 functions in this file to start/stop/check status containerd.

File: nodelet/tooling/tree/usr/local/lib/systemd/system/containerd.service [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+Restart=always
+RestartSec=5
+# Having non-zero Limit*s causes performance problems due to accounting overhead
+# in the kernel. We recommend using cgroups to do container-local accounting.
+LimitNPROC=infinity
+LimitCORE=infinity
+LimitNOFILE=infinity
+# Comment TasksMax if your systemd version does not supports it.
+# Only systemd 226 and above support this version.
+TasksMax=infinity
```
Comment (pacharya-pf9) [inline nodelet/tooling/tree/usr/local/lib/systemd/system/containerd.service:22]: Systemd on CentOS is 219 only

File: nodelet/pkg/phases/container_runtime/configure_containerd.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+			}
+		} else {
+			appendata = fmt.Sprint("\n[plugins.\"io.containerd.grpc.v1.cri\".containerd.runtimes.runc.options]\n\t\tSystemdCgroup = false")
+			err = file.WriteToFile(constants.ContainerdConfigFile, appendata, true)
+			if err != nil {
+				phaseutils.SetHostStatus(cp.hostPhase, constants.FailedState, "couldn't write to containerd config file")
+				return errors.Wrapf(err, "couldn't write to containerd config file:%s", constants.ContainerdConfigFile)
+			}
+		}
+	}
```
Comment (pacharya-pf9) [inline nodelet/pkg/phases/container_runtime/configure_containerd.go:98]:
```suggestion
		var appendData := fmt.Sprint("\n[plugins.\"io.containerd.grpc.v1.cri\".containerd.runtimes.runc.options]\n\t\tSystemdCgroup = false")
		if constants.ContainerdCgroup == "systemd" {
			appendData = fmt.Sprint("\n\t[plugins.\"io.containerd.grpc.v1.cri\".containerd.runtimes.runc.options]\n\t\tSystemdCgroup = true")
		}
	        err = file.WriteToFile(constants.ContainerdConfigFile, appendData, true)
		if err != nil {
			phaseutils.SetHostStatus(cp.hostPhase, constants.FailedState, "couldn't write to containerd config file")
			return errors.Wrapf(err, "couldn't write to containerd config file:%s", constants.ContainerdConfigFile)
		}
	}
```

**Review by pacharya-pf9 — APPROVED** [review]

**Review by pacharya-pf9 — APPROVED** [review]

**Comment by xagent003** [comment]
I rebased the branch to main since it was a bit out of date. Build a new nodelet RPM, see this error:

```
[root@arjunairc7803 ~]# tail -f /var/log/pf9/nodelet.log 
{"L":"WARN","T":"2022-09-09T21:15:22.693Z","C":"nodelet/nodelet.go:410","M":"Sunpike communication is disabled; not sending an update to sunpike-conductor."}
{"L":"WARN","T":"2022-09-09T21:15:22.694Z","C":"nodelet/nodelet.go:410","M":"Sunpike communication is disabled; not sending an update to sunpike-conductor."}
{"L":"INFO","T":"2022-09-09T21:15:22.694Z","C":"container_runtime/start_containerd.go:52","M":"Running Start of phase: Start Container Runtime"}
{"L":"INFO","T":"2022-09-09T21:15:22.694Z","C":"command/check_service.go:29","M":"Running containerd start"}
{"L":"INFO","T":"2022-09-09T21:15:22.694Z","C":"command/command.go:106","M":"Running command '/bin/sudo /usr/bin/systemctl start containerd' from wd: ''"}
{"L":"INFO","T":"2022-09-09T21:15:22.694Z","C":"command/command.go:117","M":"STDOUT: "}
{"L":"INFO","T":"2022-09-09T21:15:22.738Z","C":"command/check_service.go:34","M":"containerd start succeeded"}
{"L":"WARN","T":"2022-09-09T21:15:22.739Z","C":"nodelet/nodelet.go:410","M":"Sunpike communication is disabled; not sending an update to sunpike-conductor."}
{"L":"WARN","T":"2022-09-09T21:15:22.740Z","C":"nodelet/nodelet.go:410","M":"Sunpike communication is disabled; not sending an update to sunpike-conductor."}
{"L":"INFO","T":"2022-09-09T21:15:22.740Z","C":"container_runtime/load_images.go:78","M":"Running Start of phase: Load user images to container runtime"}
{"L":"INFO","T":"2022-09-09T21:16:55.872Z","C":"container_runtime/image_utils.go:47","M":"Loading images from file: /var/opt/pf9/images/docker-imgs-v-5.6.0-2161634.tar.gz"}
{"L":"ERROR","T":"2022-09-09T21:16:55.892Z","C":"container_runtime/load_images.go:106","M":"could not load images from : /var/opt/pf9/images/docker-imgs-v-5.6.0-2161634.tar.gz: failed to create container runtime client: failed to dial \"/run/containerd/containerd.sock\": connection error: desc = \"transport: error while dialing: dial unix /run/containerd/containerd.sock: connect: permission denied\""}
```

**Comment by pacharya-pf9** [comment]
@srm6867 
The image upload phase fails with this change - 
```
{"L":"WARN","T":"2022-09-09T21:15:22.693Z","C":"nodelet/nodelet.go:410","M":"Sunpike communication is disabled; not sending an update to sunpike-conductor."}
{"L":"WARN","T":"2022-09-09T21:15:22.694Z","C":"nodelet/nodelet.go:410","M":"Sunpike communication is disabled; not sending an update to sunpike-conductor."}
{"L":"INFO","T":"2022-09-09T21:15:22.694Z","C":"container_runtime/start_containerd.go:52","M":"Running Start of phase: Start Container Runtime"}
{"L":"INFO","T":"2022-09-09T21:15:22.694Z","C":"command/check_service.go:29","M":"Running containerd start"}
{"L":"INFO","T":"2022-09-09T21:15:22.694Z","C":"command/command.go:106","M":"Running command '/bin/sudo /usr/bin/systemctl start containerd' from wd: ''"}
{"L":"INFO","T":"2022-09-09T21:15:22.694Z","C":"command/command.go:117","M":"STDOUT: "}
{"L":"INFO","T":"2022-09-09T21:15:22.738Z","C":"command/check_service.go:34","M":"containerd start succeeded"}
{"L":"WARN","T":"2022-09-09T21:15:22.739Z","C":"nodelet/nodelet.go:410","M":"Sunpike communication is disabled; not sending an update to sunpike-conductor."}
{"L":"WARN","T":"2022-09-09T21:15:22.740Z","C":"nodelet/nodelet.go:410","M":"Sunpike communication is disabled; not sending an update to sunpike-conductor."}
{"L":"INFO","T":"2022-09-09T21:15:22.740Z","C":"container_runtime/load_images.go:78","M":"Running Start of phase: Load user images to container runtime"}
{"L":"INFO","T":"2022-09-09T21:16:55.872Z","C":"container_runtime/image_utils.go:47","M":"Loading images from file: /var/opt/pf9/images/docker-imgs-v-5.6.0-2161634.tar.gz"}
{"L":"ERROR","T":"2022-09-09T21:16:55.892Z","C":"container_runtime/load_images.go:106","M":"could not load images from : /var/opt/pf9/images/docker-imgs-v-5.6.0-2161634.tar.gz: failed to create container runtime client: failed to dial \"/run/containerd/containerd.sock\": connection error: desc = \"transport: error while dialing: dial unix /run/containerd/containerd.sock: connect: permission denied\""}
```
Can you please take a look? Consider changing the file permissions of the containerd binary or the socket file that it creates to pf9 user so that nodelet can access it.

**Comment by xagent003** [comment]
Other issue is that /var/lib/containerd folder(and any others) is not created. Perhaps this was handled by RPM install. After chowning containerd.sock, it errored out with:

```
{"L":"INFO","T":"2022-09-09T21:38:04.056Z","C":"command/command.go:106","M":"Running command '/bin/sudo /usr/bin/systemctl start containerd' from wd: ''"}
{"L":"INFO","T":"2022-09-09T21:38:04.057Z","C":"command/command.go:117","M":"STDOUT: "}
{"L":"INFO","T":"2022-09-09T21:38:04.102Z","C":"command/check_service.go:34","M":"containerd start succeeded"}
{"L":"WARN","T":"2022-09-09T21:38:04.103Z","C":"nodelet/nodelet.go:410","M":"Sunpike communication is disabled; not sending an update to sunpike-conductor."}
{"L":"WARN","T":"2022-09-09T21:38:04.104Z","C":"nodelet/nodelet.go:410","M":"Sunpike communication is disabled; not sending an update to sunpike-conductor."}
{"L":"INFO","T":"2022-09-09T21:38:04.104Z","C":"container_runtime/load_images.go:78","M":"Running Start of phase: Load user images to container runtime"}
{"L":"INFO","T":"2022-09-09T21:39:00.143Z","C":"container_runtime/image_utils.go:47","M":"Loading images from file: /var/opt/pf9/images/kubedu-imgs-v-5.6.0-2161634.tar.gz"}
{"L":"ERROR","T":"2022-09-09T21:39:00.232Z","C":"container_runtime/load_images.go:106","M":"could not load images from : /var/opt/pf9/images/kubedu-imgs-v-5.6.0-2161634.tar.gz: failed to import images from: /var/opt/pf9/images/kubedu-imgs-v-5.6.0-2161634.tar.gz: failed to ingest \"00a6555ada730bdd3097d229383e0df3f20b3c0e841b2962b50c440d8667ca32/VERSION\": failed to open writer: mkdir /var/lib/containerd/io.containerd.content.v1.content/ingest/d3af7cd621aa7dd23c31558ca803dcab371e3632f3e4290005c609def1849b4c: no such file or directory: unknown"}
```

I normally yum erase containerd and rm -rf /var/lib/containerd and other related folders when re-using hosts to start fresh, and it has worked before.

**Comment by xagent003** [comment]
Hold off. Containerd is installed and images loaded, I see etcd and kube-proxy running. But cluster failed on waiting for apiserver to run. Let me debug to rule out it's caused by this commit, as it was working last time I deployed without this change

**Comment by xagent003** [comment]
kubelet cant start apiserver with these changes.

**Pull request merged**